### PR TITLE
Update: Add ignorePattern to no-inline-comments

### PR DESCRIPTION
**Pull request by EdieLemoine**
Update: Add filtering to no-inline-comments and allow Webpack magic comments 

#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).
- [ ] The team has reached consensus on the changes proposed in this pull request. If not, I understand that the evaluation process will begin with this pull request and won't be merged until the team has reached consensus.

#### What is the purpose of this pull request?

- [ ] Documentation update
- [ ] Bug fix
- [ ] New rule 
- [X] Changes an existing rule
- [ ] Add autofixing to a rule
- [ ] Add a CLI option
- [ ] Add something to the core
- [ ] Other, please explain:

**What rule do you want to change?**
no-inline-comments 

**Does this change cause the rule to produce more or fewer warnings?**
Fewer

**How will the change be implemented?**
New default behaviour. 

**Please provide some example code that this change will affect:**

```js
const language = import(/* webpackChunkName: "my-chunk-name" */'./locale/language');
```

**What does the rule currently do for this code?**
Reports error/warning: `Unexpected comment inline with code.(no-inline-comments)`

**What will the rule do after it's changed?**
Not report on this kind of comment anymore.

#### What changes did you make?
I added all the [Webpack magic comments](https://webpack.js.org/api/module-methods/#magic-comments] keywords to the `astUtils.COMMENTS_IGNORE_PATTERN` regex. It looks like this now:

```js
// From this:
/^\s*(?:eslint|jshint\s+|jslint\s+|istanbul\s+|globals?\s+|exported\s+|jscs)/u
// To this:
/^\s*(?:eslint|jshint\s+|jslint\s+|istanbul\s+|globals?\s+|exported\s+|jscs|webpackInclude:|webpackExclude:|webpackChunkName:|webpackMode:|webpackPrefetch:|webpackPreload:)/u
```
This affects the following rules:
* no-inline-comments
* capitalized-comments (+ This is a big plus as it would break the keywords)
* line-comment-position (- Likely not necessary)
* lines-around-comment (- Likely not necessary)
* multiline-comment-style (- Likely not necessary)

I added a filter to the `sourceCode.getAllComments()` call in the `Program` method of the rule's definition so these comments (and the ones that were already present in `COMMENTS_IGNORE_PATTERN`) will be ignored.

#### Is there anything you'd like reviewers to focus on?
Is the `astUtils.COMMENTS_IGNORE_PATTERN` is the correct place for the new exceptions? If so, I will update the docs in every place it applies.
Should this be the new default behaviour?

### CODE REVIEW

**Comment by kaicataldo** [comment]
Thanks for the PR! While I think this is a use-case that would be good to support, I personally don't think this is the correct solution. The current ignored comments are all either ESLint directive comments or comments for other linters (I believe this is to ease the pain of switching to ESLint from `jshint`, `jscs`, etc.).

I would advocate for adding a new option to `no-inline-comments` that allows you to specify an array of exceptions. Maybe `ignorePattern` to match the option in [`capitalized-comments`](https://eslint.org/docs/rules/capitalized-comments#ignorepattern)?

Please note that we'll need to go through the consensus process outlined [here](https://eslint.org/docs/maintainer-guide/issues#consensus) before we can merge anything. Feel free to wait implementing anything else until we come to a decision so that you don't have to spend more time on it!

@eslint/eslint-team Thoughts?

**Comment by mdjermanovic** [comment]
`line-comment-position`, `capitalized-comments` and `lines-around-comment` rules have `ignorePattern` option. 

This option already allows users to define custom patterns like those for webpack magic comments, so I don't think we should change the default pattern (`astUtils.COMMENTS_IGNORE_PATTERN`) which affects all these rules. It looks like an unnecessary breaking change.

As for the `no-inline-comments` rule, I believe it should have the same `ignorePattern` option (maybe `applyDefaultIgnorePatterns` as well). There is already an open issue #12755 for this enhancement.

**Comment by kaicataldo** [comment]
@EdieLemoine Would you be interested in helping us implement the suggestion above once we come to a consensus on its inclusion?

**Comment by EdieLemoine** [comment]
@kaicataldo sure. Just let me know which solution you guys prefer.

**Comment by kaicataldo** [comment]
I will champion this change (as described [here](https://github.com/eslint/eslint/pull/13029#issuecomment-601504598)). Given the two :+1: on my proposal, we only need one more :+1:.

**Comment by kaicataldo** [comment]
And thanks @EdieLemoine! Once we reach consensus we'll make sure it's clear here (hopefully soon since we only need one more :+1:).

**Comment by kaicataldo** [comment]
@eslint/eslint-team We just need one more :+1: to accept this.

**Comment by nzakas** [comment]
I've added my :+1: so marking as accepted.

**Comment by mdjermanovic** [comment]
Linked #12755 as it seems that this option will fix the issue with `istanbul` comments, too.

**Comment by mdjermanovic** [comment]
@kaicataldo

> I would advocate for adding a new option to `no-inline-comments` that allows you to specify an array of exceptions. Maybe `ignorePattern` to match the option in [`capitalized-comments`](https://eslint.org/docs/rules/capitalized-comments#ignorepattern)?

to clarify, should the option accept an _array_ of "regex" strings or just one string?

The same option in similar rules like `line-comment-position`, `capitalized-comments` and `lines-around-comment` accepts 1 string, which provides basically the same functionality (maybe just a bit less convenient in some cases).

**Comment by kaicataldo** [comment]
My take is that we should math what those other rules are currently doing. Supporting RegExp strings would be an additive change, and we can add it to all the rules listed if we decide we want to go that route in the future.

**Comment by kaicataldo** [comment]
@EdieLemoine Thanks for your patience! Is this still something you're willing to help out with?

**Comment by EdieLemoine** [comment]
@kaicataldo I just updated it. Is this the proper solution?

File: lib/rules/no-inline-comments.js
Comment (mdjermanovic) [inline lib/rules/no-inline-comments.js:109]: This shouldn't be included in this change.

File: tests/lib/rules/no-inline-comments.js
Comment (mdjermanovic) [inline tests/lib/rules/no-inline-comments.js:99]: This example would be valid even without the pattern, since the comments are on separate lines?

File: tests/lib/rules/no-inline-comments.js
Comment (mdjermanovic) [inline tests/lib/rules/no-inline-comments.js:107]: Can we also add some `invalid` examples where the comment doesn't match the configured pattern

File: tests/lib/rules/no-inline-comments.js
Comment (EdieLemoine) [reply tests/lib/rules/no-inline-comments.js:99]: Now that's silly 🤦‍♀️

**Comment by EdieLemoine** [comment]
@mdjermanovic I updated the branch.

File: docs/rules/no-inline-comments.md
Comment (mdjermanovic) [inline docs/rules/no-inline-comments.md:46]: Due to recent changes in `levn`, which we use to parse config values in comments, this should be `\\s` like in string literals:

```js
/*eslint no-inline-comments: ["error", { "ignorePattern": "webpackChunkName:\\s.+" }]*/
```

**Comment by EdieLemoine** [comment]
@mdjermanovic I updated it again :)

**Review by kaicataldo — DISMISSED** [review]
This looks really great - thanks! One small comment about clarifying the docs, but otherwise LGTM.

File: docs/rules/no-inline-comments.md
Comment (EdieLemoine) [reply docs/rules/no-inline-comments.md:95]: Hmm, it's implemented the same way as in lines-around-comment. It's referenced to as a "regular expression" there as well.

Comment (btmills) [reply docs/rules/no-inline-comments.md:95]: That's a good point, @EdieLemoine. Since that's the only change @kaicataldo requested, let's merge this to get it into today's release, and I'll put together a quick PR to change this in both places.

**Review by btmills — APPROVED** [review]
Thanks for contributing, @EdieLemoine! I'll update the docs after merging, and this will be in today's release.

**Pull request merged**
